enhancement/1732 - Add pronunciation to fixes in airport files

OlaRune opened the pull request:
Resolves #1732

The purpose of this pull request is to add an optional `spoken` parameter to fixes that defines what is said by the speech engine.

OlaRune commented on src/assets/scripts/client/navigationLibrary/FixCollection.js:
```diff
@@ -162,8 +180,11 @@ class FixCollection extends BaseCollection {
-            const fixModel = new FixModel(fixName, fixCoordinates, referencePosition);
+        _forEach(fixList, (fixData) => {
+            if (!fixData.spoken) {
+                fixData.spoken = null; // the initializer for FixModel handles this correctly
```
Maybe this is unnecessary. Is it already `null` or `undefined`? (comment for myself)

OlaRune replied:
Most likely `undefined` if it is not in the JSON. I have changed it and tested and it seems to work.

OlaRune commented on src/assets/scripts/client/navigationLibrary/FixCollection.js:
```diff
@@ -100,6 +100,24 @@ class FixCollection extends BaseCollection {
+       const fixModel = this.findFixByName(fixName);
+
+       if (!fixModel) {
+           return fixName.toLowerCase(); // lower case to force speech engine to pronounce instead of spell
```
Maybe this should be upper-case? How are strange fix names pronounced in lower-case? (comment for myself)

OlaRune replied:
I've experimented a bit, and it will spell fixes that are completely impossible to pronounce, for example `XFDCG`. However, other names, such as `ASDFG`, will have it try to pronounce them, which sounds super weird. I think I'll have to make it spell out all fixes that don't have a defined `spoken`.

@Fechulo: See this comment as the answer to our discussion on Slack. Too bad it didn't work out nicely! :sob:

OlaRune replied:
This is now changed.

OlaRune commented on src/assets/scripts/client/navigationLibrary/FixModel.js:
```diff
@@ -74,23 +84,41 @@ export default class FixModel extends BaseModel {
+
+        if (!fixSpoken) {
+            this._spoken = fixName.toLowerCase();
+        } else {
```
Assign the default value before the `if` to avoid the `else`. (comment for myself)

OlaRune replied:
This is fixed.

OlaRune commented on src/assets/scripts/client/navigationLibrary/FixModel.js:
```diff
@@ -74,23 +84,40 @@ export default class FixModel extends BaseModel {
         }
 
         this.name = fixName.toUpperCase();
+        this._spoken = this.name; // upper-case forces spelling, experiments shows that this works best for unknown pronunciation
```
I'm still not completely sure about this, I'll let you reviewers say what you think of this.

**Uppercase**
Forces names to always be spelled out if we don't have a defined `spoken` parameter. This will always lead to reasonable results, although we miss some opportunities where the voice engine could have pronounced the names in a better way.

**Lowercase**
This forces the voice engine to try pronouncing the names as words. Works nicely for names resembling words, such as `AUBRN`. Also works well for some hard-to-pronounce names, such as  `XDRCX`, which it spells out instead of pronouncing. Produces really weird results for other names that shouldn't be pronounced but it still tries, such as `ASDFG`, which makes it impossible to understand what it is saying.

What do you think? The lowercase version was used before, but I think it's better to force spelling now, since we can override it by defining a `spoken` instead.

erikquinn replied:
Definitely still lowercase. Most waypoints have some sort of discernable pronunciation, and that is what would be used on frequency, not spelling out a 5-letter fix. And as you mention, when the fix doesn't have a reasonable pronunciation, it spells it out. But we definitely don't want to require a spoken parameter to prevent the sim from pronouncing `APPLE` as `alpha papa papa lima echo`-- the `spoken` parameter is the exception when we want it to be pronounced a specific way which is different than what the robot lady chose :)

OlaRune replied:
Alright, I'll change it back to lowercase so we'll let the airport developer make sure they are pronounced correctly if something is weird.

OlaRune commented:
**For the reviewers:**
I have added `spoken` parameters to 3 fixes at KSEA, since those are used in `airport-format.md`. These can be used if you want to test the new functionality from the PR deployment.

For example, order a `direct` to any of these to hear it.
- EUG
- KRUZR
- PAE

erikquinn commented on assets/airports/cyhz.json:
```diff
@@ -588,96 +588,96 @@
-        "ZHZ"  : ["N44d48.14", "W063d35.38"],
-        "ZNS"  : ["N44d58.00", "W063d25.62"]
-    },
+    "fixes": [
+        {"name": "AGMIR", "coordinates": ["N45d13.70", "W061d57.33"]},
+        {"name": "ALLEX", "coordinates": ["N44d25.00", "W067d00.00"]},
+        {"name": "AMUNA", "coordinates": ["N44d57.74", "W063d07.54"]},
+        {"name": "ARDEE", "coordinates": ["N45d07.40", "W064d23.17"]},
```
This pattern worries me. If using an array of objects rather than the fix name as the key of an object property, we introduce the possibility of duplicated fixes. As it stands currently, this is protected by a lint error (attempting to define two of the same key in an object). I would suggest instead something like:
```
"fixes": {
        "AGMIR": ["N45d13.70", "W061d57.33"],
        "ALLEX": ["N44d25.00", "W067d00.00"],
        "AMUNA": {"coordinates": ["N44d57.74", "W063d07.54"], "spoken": "amoonuh"},
        "ARDEE", {"coordinates": ["N45d07.40", "W064d23.17"], "spoken": "party},
}
```

and then from the FixModel constructor, we can simply check if `Array.isArray(fixData)`-- if so, treat as positional array with spoken undefined, and else `const { coordinates, spoken } = fixData;` to pull out the two values and use them. If the fix data provided is not an object with those two properties, an error will be thrown naturally, ensuring the existence of the parameters without any other manual defensiveness. 😄

OlaRune replied:
Wilco! That sounds great.

erikquinn commented on src/assets/scripts/client/navigationLibrary/FixCollection.js:
```diff
@@ -162,8 +180,8 @@ class FixCollection extends BaseCollection {
-        _forEach(fixList, (fixCoordinates, fixName) => {
-            const fixModel = new FixModel(fixName, fixCoordinates, referencePosition);
+        _forEach(fixList, (fixData) => {
+            const fixModel = new FixModel(fixData.name, fixData.spoken, fixData.coordinates, referencePosition); // FixModel accepts undefined spoken
```
We should just pass `fixData` here, and have the FixModel constructor destruct the object in its _init() method. I can see you were just matching the pattern that was already there, but where practical we prefer to squish these together into an object and pass that object to the constructor. If anything is missing, we deal with it in ._init() and throw/return as appropriate.

OlaRune replied:
Sounds good!

erikquinn commented on src/assets/scripts/client/navigationLibrary/FixCollection.js:
```diff
@@ -100,6 +100,24 @@ class FixCollection extends BaseCollection {
+    /**
+     * Returns the pronunciation of the fix name if it exists, otherwise returns the input name
+     *
+     * @for FixCollection
+     * @method getFixSpokenName
+     * @param fixName {string}
+     * @return {string}
+     */
+    getFixSpokenName(fixName) {
```
I find it hard to think of a case where we would have the fix name, but not the FixModel itself. Should we not just achieve this with `FixModel.getSpokenName()`?

Is this maybe just for the readbacks, like with the `pd` command? If so and the fix model isn't readily accessible (or inconvenient), this is fine. Just checking since I wasn't expecting to see it here.

OlaRune replied:
There are two reasons why I chose to do it this way.

1. Easier use of the function for evaluating the spoken name directly in readback strings, without having to look up if the actual fix exists or not. If it doesn't exist the input is read back.

2. The readback string is constructed in a few different places, with access to different object types. I'm on mobile right now, but I recall that somewhere it had a FixModel, in most places a WaypointModel, and I think there was a HoldModel or something at one place. These other types of models don't have an easy way to access the underlying FixModel, they have just copied the fix name. So we can't access the FixModel without searching for it in the FixCollection anyway. So I decided this solution would need less code and lookups from different places.

I prefer to leave this as it is, if you don't have very strong feelings about doing it another way.

erikquinn replied:
Okay, that all makes sense. Fine as is.

OlaRune commented:
@erikquinn I have updated the pull request according to your comments. Since it is now backwards compatible with the old fix definition syntax I have reverted the changes to the airport files. Also updated tests accordingly and added a few new tests.

Feel free to review again when you have time. Less files to look through now. :smile:

erikquinn reviewed (approved):
Pending lint/tests passing, this is approved with a slight modification.

Not sure why I didn't suggest this instead in the first place, but there's really no need for an object-based definition at all. lat/lon are always required, and the spoken parameter is always optional... so we can just stick with the array shape, and where desired, add the spoken string as a third element in the array. This also reduces a lot of the required defensiveness involved with handling different ways the input data could be malformed by new airport devs 😉 

Now simply works with:
```javascript
"FIXXA": [41.124, -81.822],
"FIXXB": [41.922, -82.049, "fixbee"],
"FIXXC": [40.732, -81.527, "yabba dabba doo"],
"FIXXD": [40.284, -81.181],
...
```
